EducationalAssignments/SecureHashPartOne.md

```diff
@@ -0,0 +1,313 @@
+version, you need to use the Windows command line.  For Windows 7 this is
+PowerShell.  You can open a new terminal by going to start, search, type
+powershell.  If you downloaded the Linux version you must use a Linux OS
+and Linux terminal.  
+
+-->
+
+
+### Tutorials for Repy and Python
+----
+Now that you have Repy and Python, you may need a refresher on how to use
+them.  The following tutorials provide this information.
+
+ * Official [Python tutorial](http://docs.python.org/tutorial/)
+ * [Differences between RepyV2 and Python](../Programming/PythonVsRepyV2.md)
+ * List of [RepyV2 API calls](../Programming/RepyV2API.md)
+
+
+
+## Building the security layer
+----
+The following program is a basic and incomplete sample code for you to get
+an idea about writing security layer. Remember, you have no idea how the
+attacker will try to penetrate your security layer, so it is important that
+you leave nothing to chance!  
+
+
+### A basic (and inadequate) defense
+
+Time to start coding!  Let's inspect a basic security layer.  
+
+```
+"""
+This security layer inadequately handles secured hashing for files in RepyV2.
+
+
+
+Note:
+    This security layer uses encasementlib.r2py, restrictions.default, repy.py and Python
+    Also you need to give it an application to run.
+    python repy.py restrictions.default encasementlib.r2py [security_layer].r2py [attack_program].r2py 
+    
+    """ 
+TYPE="type"
+ARGS="args"
+RETURN="return"
+EXCP="exceptions"
+TARGET="target"
+FUNC="func"
+OBJC="objc"
+
+class SecureFile():
+	def __init__(self,filename,create):
+		# globals
+		mycontext['debug'] = False   
+		
+		# local (per object) reference to the underlying file
+		self.fn = filename
+		self.trackfn = 'trackhash'
+		
+		# make the files
+		if create:
+			# ensure to check your filename
+			self.file = openfile(self.fn,create)
+			self.trackfile = openfile(self.trackfn,True)
+
+	def writeat(self,data,offset):	
+		# applying hashfunction to your data - repy supports SHA1 library
+		temp_data = applyHashFunction(data)
+    		block_no = 0
+		
+		# Writing file contents in the "trackhash" file with filename, block no and hashed data
+		# your data may not necessarily be in one entire block or begin from block no 0
+		self.trackfile.writeat(self.fn+" "+str(block_no)+" "+temp_data,0)
+		
+		#close the trackhash file
+		self.trackfile.close()
+
+		# Write the requested data to the file using the sandbox's writeat call
+		self.file.writeat(data, offset)		
+  
+	def readat(self,bytes,offset):
+		# Read from the file using the sandbox's readat...
+		# read the file only if the block is first written to
+		return self.file.readat(bytes,offset)
+
+	def close(self):
+		self.file.close()
+
+def secureopenfile(filename, create):
+	return SecureFile(filename,create)
+
+# The code here sets up type checking and variable hiding for you.  You
+# should not need to change anything below here.
+sec_file_def = {"obj-type":SecureFile,
```

Comment:
Don't you need to pass up the call for count_readat()?

Reply:
oh yes, I will just add it